File: scripts/update.ts
#!/usr/bin/env bun
/** GoodBehavior deterministic updater — the mechanical half of /update-goodbehavior.
 * Reads manifest, resolves source repo, reconciles tracked files via git 3-way merge.
 * Skill keeps JUDGMENT; this script does MERGE MECHANICS identically every run.
 *
 * Usage: bun scripts/update.ts --target /path/to/project [--source /override] [--dry-run]
 *
 * Guarantees: never touches untracked files; untouched files fast-forward; adapted files 3-way merged;
 * conflicts written with markers; upstream-removed files kept locally.
 * Reports JSON: {"status": "...", "unchanged": [], "updated": [], "merged": [], "conflict": [],
 *   "restored": [], "removed_upstream": [], "warnings": []}
 * Exit: 0 clean, 1 fatal, 2 conflicts.
 */

import { parseArgs } from "node:util";
import { createHash } from "node:crypto";
import { readFileSync, writeFileSync, mkdirSync, existsSync, rmSync } from "node:fs";
import { join, resolve, dirname } from "node:path";
import { execFileSync } from "node:child_process";
import { tmpdir } from "node:os";

interface Manifest {
  source: string;
  sourceCommit: string | null;
  installedAt: string;
  updatedAt: string | null;
  files: Record<string, { from: string; sha256: string }>;
}

interface Report {
  status: string;
  unchanged: string[];
  updated: string[];
  merged: string[];
  conflict: string[];
  restored: string[];
  removed_upstream: string[];
  warnings: string[];
}

function sha256Bytes(b: Uint8Array): string {
  return createHash("sha256").update(b).digest("hex");
}

function git(source: string, ...args: string[]): Buffer | null {
  try { return execFileSync("git", ["-C", source, ...args], { stdio: "pipe" }); }
  catch { return null; }
}

function gitShow(source: string, commit: string, path: string): Uint8Array | null {
  const out = git(source, "show", `${commit}:${path}`);
  return out ? new Uint8Array(out) : null;
}

function mergeFile(ours: Uint8Array, base: Uint8Array, theirs: Uint8Array): [Uint8Array, boolean] {
  const d = join(tmpdir(), `gbu-${Date.now()}-${Math.random().toString(36).slice(2)}`);
  mkdirSync(d);
  try {
    const [po, pb, pt] = ["ours", "base", "theirs"].map(n => join(d, n));
    [po, pb, pt].forEach((p, i) => writeFileSync(p, [ours, base, theirs][i]));
    const out = execFileSync("git",
      ["merge-file", "-p", "-L", "local", "-L", "base", "-L", "upstream", po, pb, pt],
      { stdio: "pipe", maxBuffer: 64 * 1024 * 1024 });
    return [new Uint8Array(out), true];
  } catch (e: unknown) {
    const stdout = e && typeof e === "object" && "stdout" in e ? (e as Record<string, unknown>).stdout : null;
    return [new Uint8Array(Buffer.isBuffer(stdout) ? stdout : Buffer.alloc(0)), false];
  } finally { rmSync(d, { recursive: true, force: true }); }
}

function main(): number {
  const { values } = parseArgs({
    args: Bun.argv.slice(2),
    options: { target: { type: "string" }, source: { type: "string" }, "dry-run": { type: "boolean" } },
    strict: true,
    allowPositionals: true,
  });

  if (!values.target) { console.error("error: --target is required"); return 1; }
  const target = resolve(values.target);
  const sourceOverride = values.source ? resolve(values.source) : null;
  const dryRun = values["dry-run"] ?? false;

  const manifestPath = join(target, ".omp", "goodbehavior", "manifest.json");
  if (!existsSync(manifestPath)) {
    console.log(JSON.stringify({ status: "no manifest — run /adopt-goodbehavior first" }, null, 2));
    return 1;
  }

  const manifest = JSON.parse(readFileSync(manifestPath, "utf8")) as Manifest;
  const source = resolve(sourceOverride ?? manifest.source);
  if (!existsSync(join(source, ".git"))) {
    console.log(JSON.stringify({ status: `source is not a git repo: ${source}` }, null, 2));
    return 1;
  }
  const baseCommit = manifest.sourceCommit;
  if (!baseCommit) {
    console.log(JSON.stringify({ status: "manifest sourceCommit is null — no merge base; re-adopt or overwrite manually" }, null, 2));
    return 1;
  }
  const newCommit = (git(source, "rev-parse", "HEAD")?.toString("utf8") ?? "").trim();
  if (!newCommit) { console.log(JSON.stringify({ status: "could not resolve source HEAD" }, null, 2)); return 1; }
  if (newCommit === baseCommit) { console.log(JSON.stringify({ status: "already up to date", commit: newCommit }, null, 2)); return 0; }

  const report: Report = { status: `${baseCommit.slice(0, 8)} -> ${newCommit.slice(0, 8)}`, unchanged: [], updated: [], merged: [], conflict: [], restored: [], removed_upstream: [], warnings: [] };

  for (const [key, entry] of Object.entries(manifest.files).sort(([a], [b]) => a.localeCompare(b))) {
    const theirs = gitShow(source, newCommit, entry.from);
    if (!theirs) { report.removed_upstream.push(key); continue; }

    const localPath = join(target, key);
    if (!existsSync(localPath)) {
      if (!dryRun) {
        mkdirSync(dirname(localPath), { recursive: true });
        writeFileSync(localPath, theirs);
        entry.sha256 = sha256Bytes(theirs);
      }
      report.restored.push(key);
      continue;
    }

    const ours = new Uint8Array(readFileSync(localPath));
    if (Buffer.compare(ours, theirs) === 0) {
      entry.sha256 = sha256Bytes(ours);
      report.unchanged.push(key);
      continue;
    }
    if (sha256Bytes(ours) === entry.sha256) {
      if (!dryRun) {
        writeFileSync(localPath, theirs);
        entry.sha256 = sha256Bytes(theirs);
      }
      report.updated.push(key);
      continue;
    }

    const base = gitShow(source, baseCommit, entry.from);
    if (!base) { report.warnings.push(`${key}: no base at ${baseCommit.slice(0, 8)} — left as-is`); continue; }

    const [merged, clean] = mergeFile(ours, base, theirs);
    if (!dryRun) {
      writeFileSync(localPath, merged);
      if (clean) entry.sha256 = sha256Bytes(merged);
    }
    report[clean ? "merged" : "conflict"].push(key);
  }

  if (!dryRun) {
    manifest.sourceCommit = newCommit;
    manifest.updatedAt = new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
    writeFileSync(manifestPath, JSON.stringify(manifest, null, 2) + "\n");
  }

  console.log(JSON.stringify(report, null, 2));
  return report.conflict.length ? 2 : 0;
}

process.exit(main());

File: scripts/update.py
#!/usr/bin/env python3
"""GoodBehavior deterministic updater — the mechanical half of /update-goodbehavior.

Reads the target's manifest, resolves the source repo, and reconciles every tracked file with a
git 3-way merge: base = the commit recorded in the manifest, ours = the local file, theirs = the
new upstream version. Untouched files fast-forward; adapted files merge; genuine conflicts get
markers for the human. The skill keeps the JUDGMENT (deciding about upstream additions/removals,
resolving conflicts with the user); this script does the merge mechanics identically every run.

Usage:
    python3 scripts/update.py --target /path/to/project [--source /override/path] [--dry-run]

Guarantees:
  - never touches anything outside its tracked files (project-owned files stay untouched)
  - a file whose sha256 matches the manifest (untouched since install) is fast-forwarded
  - an adapted file is 3-way merged; conflicts are written WITH markers and reported, and their
    manifest sha256 is left stale so a re-run after resolution treats them cleanly
  - files upstream REMOVED are kept locally and reported (deleting is the skill/user's call)
  - manifest sourceCommit/updatedAt/sha256s refreshed for everything cleanly updated or merged
Reports JSON on stdout:
  {"status": "...", "unchanged": [], "updated": [], "merged": [], "conflict": [],
   "restored": [], "removed_upstream": [], "warnings": []}
Exit code: 0 clean (even if already up to date), 1 fatal, 2 completed WITH conflicts.
"""
import argparse, hashlib, json, os, subprocess, sys, tempfile
from datetime import datetime, timezone


def sha256_bytes(b):
    return hashlib.sha256(b).hexdigest()


def git(source, *args, binary=False):
    out = subprocess.run(["git", "-C", source] + list(args), capture_output=True)
    if out.returncode != 0:
        return None
    return out.stdout if binary else out.stdout.decode()


def git_show(source, commit, path):
    return git(source, "show", f"{commit}:{path}", binary=True)


def merge_file(ours, base, theirs):
    """git merge-file on byte content. Returns (merged_bytes, clean)."""
    with tempfile.TemporaryDirectory() as d:
        po, pb, pt = (os.path.join(d, n) for n in ("ours", "base", "theirs"))
        for p, content in ((po, ours), (pb, base), (pt, theirs)):
            with open(p, "wb") as f:
                f.write(content)
        r = subprocess.run(["git", "merge-file", "-p", "-L", "local", "-L", "base", "-L", "upstream",
                            po, pb, pt], capture_output=True)
        return r.stdout, r.returncode == 0


def main():
    ap = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--target", required=True)
    ap.add_argument("--source", help="override the manifest's source path")
    ap.add_argument("--dry-run", action="store_true")
    args = ap.parse_args()

    target = os.path.abspath(args.target)
    manifest_path = os.path.join(target, ".omp", "goodbehavior", "manifest.json")
    if not os.path.isfile(manifest_path):
        print(json.dumps({"status": "no manifest — run /adopt-goodbehavior first"}))
        return 1
    with open(manifest_path) as f:
        manifest = json.load(f)

    source = os.path.abspath(args.source or manifest["source"])
    if not os.path.isdir(os.path.join(source, ".git")):
        print(json.dumps({"status": f"source is not a git repo: {source}"}))
        return 1
    base_commit = manifest.get("sourceCommit")
    if not base_commit:
        print(json.dumps({"status": "manifest sourceCommit is null — no merge base; "
                                    "re-adopt or overwrite manually (never silently)"}))
        return 1
    new_commit = (git(source, "rev-parse", "HEAD") or "").strip()
    if not new_commit:
        print(json.dumps({"status": "could not resolve source HEAD"}))
        return 1
    if new_commit == base_commit:
        print(json.dumps({"status": "already up to date", "commit": new_commit}))
        return 0

    report = {"status": f"{base_commit[:8]} -> {new_commit[:8]}", "unchanged": [], "updated": [],
              "merged": [], "conflict": [], "restored": [], "removed_upstream": [], "warnings": []}

    for key, entry in sorted(manifest.get("files", {}).items()):
        src_rel = entry["from"]
        theirs = git_show(source, new_commit, src_rel)
        if theirs is None:
            report["removed_upstream"].append(key)
            continue
        local_path = os.path.join(target, key)
        if not os.path.isfile(local_path):
            if not args.dry_run:
                os.makedirs(os.path.dirname(local_path), exist_ok=True)
                with open(local_path, "wb") as f:
                    f.write(theirs)
                entry["sha256"] = sha256_bytes(theirs)
            report["restored"].append(key)
            continue
        with open(local_path, "rb") as f:
            ours = f.read()
        if ours == theirs:
            entry["sha256"] = sha256_bytes(ours)
            report["unchanged"].append(key)
            continue
        if sha256_bytes(ours) == entry.get("sha256"):
            # untouched since install → fast-forward to upstream
            if not args.dry_run:
                with open(local_path, "wb") as f:
                    f.write(theirs)
                entry["sha256"] = sha256_bytes(theirs)
            report["updated"].append(key)
            continue
        base = git_show(source, base_commit, src_rel)
        if base is None:
            report["warnings"].append(f"{key}: no base at {base_commit[:8]} — left as-is; merge by hand")
            continue
        merged, clean = merge_file(ours, base, theirs)
        if not args.dry_run:
            with open(local_path, "wb") as f:
                f.write(merged)
            if clean:
                entry["sha256"] = sha256_bytes(merged)
            # on conflict: leave the stale sha256 so a post-resolution re-run reconciles cleanly
        report["merged" if clean else "conflict"].append(key)

    if not args.dry_run:
        manifest["sourceCommit"] = new_commit
        manifest["updatedAt"] = datetime.now(timezone.utc).isoformat(timespec="seconds")
        with open(manifest_path, "w") as f:
            json.dump(manifest, f, indent=2)
            f.write("\n")

    json.dump(report, sys.stdout, indent=2)
    print()
    return 2 if report["conflict"] else 0


if __name__ == "__main__":
    sys.exit(main())
